lib/assertions.js

```diff
@@ -344,9 +344,104 @@ module.exports = expect => {
           this.outdentLines();
         });
       }
-      propertyNames.forEach(propertyName => {
-        expect(subject, '[not] to have [own] property', String(propertyName));
-      });
+
+      if (expect.flags.only) {
+        if (expect.flags.not) {
+          expect.errorMode = 'bubble';
+          expect.fail(
+            'The "not" flag cannot be used together with "to only have properties".'
+          );
+        }
+        if (expect.flags.own) {
+          expect.errorMode = 'bubble';
+          expect.fail(
+            'The "own" flag cannot be used together with "to only have properties".'
+          );
+        }
+        const subjectType = expect.subjectType;
+        const subjectKeys = subjectType.getKeys(subject).filter(
+          key =>
+            utils.numericalRegExp.test(key) ||
+            typeof key === 'symbol' ||
+            // include only those keys whose value is not undefined
+            typeof subjectType.valueForKey(subject, key) !== 'undefined'
+        );
+
+        expect.withError(
+          () => {
+            expect(subjectKeys.length === propertyNames.length, 'to be true');
+            // now catch differing property names
+            const keyInValue = {};
+            propertyNames.forEach(key => {
+              keyInValue[key] = true;
+            });
+            subjectKeys.forEach(key => expect(key in keyInValue, 'to be true'));
```

Comment:
This doesn’t look safe for properties like `toString`. (To me, using the `in` operator is almost always wrong)

Reply:
Hmm, I thought this was safe - subjectKeys is generated by the type that should be taking care of it - every string key on the RHS is added to a fresh object so every one should also be enumerable.

That said I’m happy to change this :) would you prefer calling though hasOwnProperty on the prototype or an undefined check against the keyInValue object?

Reply:
The problem is that the `in` operator traverses the prototype chain and will return true for properties that are on `Object.prototype`:

```js
> 'toString' in {}
true
```

You can use that to fool your implementation to pass when it clearly shouldn't:

```
expect({ toString: 123 }, 'to only have properties', ['a']);
```

... or into rendering a bad diff:

```js
expect({ toString: 123 }, 'to only have properties', []);

expected { toString: 123 } to only have properties []

{
  toString: 123
}
```

Reply:
@papandreou thanks for talking that through with me, that can get potentially gnarly! Have made the necessary corrections - result hopefully captured by the following test: https://github.com/unexpectedjs/unexpected/pull/671/files#diff-22b57e35705626d24e33c15ccf1c72e6R347 😸